### ISSUE TITLE
Total area budget error

### ISSUE THREAD

**Issue by ricschuster** [NCC-CNC/wheretowork#288]
It seems like the `Total area budget` option causes an error now.

As an example, using the built in "South Western Ontario" project and this configuration file: 
[s3_configs.zip](https://github.com/NCC-CNC/wheretowork/files/10769770/s3_configs.zip)
and toggling on `Total area budget` like this:
![image](https://user-images.githubusercontent.com/3445092/219742425-6aafe0a4-9a3a-42c5-aea4-61cb3e1c0316.png)
causes this uninformative error:
![image](https://user-images.githubusercontent.com/3445092/219742573-87a34259-7d2b-4c3e-8ac4-b15ca0a01921.png)

I've tried with different configurations/parameters and projects and get the same result. Something seems off with the `Total area budget` option.

**Comment by DanWismer** [comment]
I will take a look

**Comment by DanWismer** [comment]
This is what gets printed:
![image](https://user-images.githubusercontent.com/29556279/220394895-392f968c-e625-4222-8cb3-47bb83f5bab2.png)

I suppose it is possible for a user to set up a scenario where no solution is possible?

**Comment by ricschuster** [comment]
I will have another test later today, but it seems odd to me that those fairly simple problems would not have a solution.

**Comment by ricschuster** [comment]
How bizarre. I can't reproduce the error anymore. Do you still get the error @DanWismer?

**Comment by DanWismer** [comment]
positing this here: https://github.com/prioritizr/prioritizr/issues/84

The odd thing is that I can't get inside the `min_shortfall_result`. It hits  https://github.com/NCC-CNC/wheretowork/blob/450e7e6815c276f829f6248a28f45e17b51c2c3b/R/server_generate_new_solution.R#L130 and then goes directly to https://github.com/NCC-CNC/wheretowork/blob/450e7e6815c276f829f6248a28f45e17b51c2c3b/R/server_generate_new_solution.R#L195

I also can't find where this is located in the code base yet `no solution found (e.g. due to problem infeasibility or time limits)`

**Comment by DanWismer** [comment]
> How bizarre. I can't reproduce the error anymore. Do you still get the error @DanWismer?

weird, I am still getting the error locally. I am wondering it it has something to do with the `future::future` call

**Comment by ricschuster** [comment]
> 

its in `solve.R` line 293 of prioritizr.

**Comment by ricschuster** [comment]
> its in `solve.R` line 293 of prioritizr.



> positing this here: [prioritizr/prioritizr#84](https://github.com/prioritizr/prioritizr/issues/84)
> 
> The odd thing is that I can't get inside the `min_shortfall_result`. It hits
> 
> https://github.com/NCC-CNC/wheretowork/blob/450e7e6815c276f829f6248a28f45e17b51c2c3b/R/server_generate_new_solution.R#L130
> 
> and then goes directly to
> https://github.com/NCC-CNC/wheretowork/blob/450e7e6815c276f829f6248a28f45e17b51c2c3b/R/server_generate_new_solution.R#L195
> 
> I also can't find where this is located in the code base yet `no solution found (e.g. due to problem infeasibility or time limits)`

its in solve.R line 293 of prioritizr.

**Comment by DanWismer** [comment]
side note, I am getting the input configuration toggled off when I load in the .yaml:
![image](https://user-images.githubusercontent.com/29556279/220421392-c2310979-61f9-434b-80c3-18f2bcdf3728.png)

I believe I know the issue. Its because in the `s3_configs.yaml` status is set to no:
![image](https://user-images.githubusercontent.com/29556279/220421700-15356876-db17-4299-ae34-c97ca853b57d.png)

The reason I never saw this is because I was testing the with a `_configs.yaml` file that didn't have that parameter captured on save. I can remove that parameter on download as there is no need in having it recorded (for now)

**Comment by ricschuster** [comment]
> side note, I am getting the input configuration toggled off when I load in the .yaml: ![image](https://user-images.githubusercontent.com/29556279/220421392-c2310979-61f9-434b-80c3-18f2bcdf3728.png)
> 
> I believe I know the issue. Its because in the `s3_configs.yaml` status is set to no: ![image](https://user-images.githubusercontent.com/29556279/220421700-15356876-db17-4299-ae34-c97ca853b57d.png)
> 
> The reason I never saw this is because I was testing the new feature with a `_configs.yaml` file that didn't have that parameter captured on save. I can remove that parameter on download as there is no need in having it recorded (for now)

Ah, that's a good one to know about. Thanks Dan!

**Comment by DanWismer** [comment]
I ran https://github.com/NCC-CNC/wheretowork/blob/450e7e6815c276f829f6248a28f45e17b51c2c3b/R/server_generate_new_solution.R#L135-L157 in a `browser()` statement and got inside the `fct_min_shortfall_result.R`.

The error is occurring here:   https://github.com/NCC-CNC/wheretowork/blob/450e7e6815c276f829f6248a28f45e17b51c2c3b/R/fct_min_shortfall_result.R#L404-L406

**Comment by DanWismer** [comment]
@ricschuster does the initial problem make sense?
![image](https://user-images.githubusercontent.com/29556279/220425339-53e33438-4f0d-416c-a208-bfc87be54147.png)

**Comment by DanWismer** [comment]
... I don't currently have a good understanding on what `cost` is doing

**Comment by DanWismer** [comment]
not sure why the `Minimum shortfall objective` is **3914.88625** when the slider shows a 55% total area budget of 7,752,25 km2

**Comment by ricschuster** [comment]
> not sure why the `Minimum shortfall objective` is **3914.88625** when the slider shows a 55% total area budget of 7,752,25 km2

Thanks! Did you set any weights?
The area multiplied by 0.505 (which all pu costs seem to be set to) does result in **3914.88625**.

**Comment by DanWismer** [comment]
when I add `s3_configs.yaml` this is what the weight settings are printing:
![image](https://user-images.githubusercontent.com/29556279/220448480-54076f46-8f1f-42f6-b881-941719b411bc.png)

For some reason the HFI is at **0!** It should be -100.

When I move the weight to a different value then back to -100, everything runs fine with the total area budget set.

The weight being set at 0 I beleive is causing the issue

**Comment by DanWismer** [comment]
disregard above .... It is indeed recording the correct weight.

Oddly, I can get a solution if I drop the HFI weight down to -96
![image](https://user-images.githubusercontent.com/29556279/220451276-1b75c2c2-bf49-4dd4-82db-80cc4ae5a9e6.png)

**Comment by DanWismer** [comment]
The `initial_problem` with the above scenario when compared to -100 HFI I see the linear constraints threshold is now at 16754
![image](https://user-images.githubusercontent.com/29556279/220455303-d55f092f-3138-409e-a287-1c3bf0e42f51.png)

@ricschuster any idea what the Linear constraints are?

**Comment by ricschuster** [comment]
`Linear constraint`: https://prioritizr.net/reference/add_linear_constraints.html

**Comment by ricschuster** [comment]
Fresh set of eyes an all:

In file `fct_min_shortfall_result.R` starting line 310 we see
```
  ## process weights with negative factors
  wn_neg_idx <- which(weight_settings$status & (weight_settings$factor < 0))
  if (length(wn_neg_idx) > 0) {
    ### add these weights to rij data
    wn_neg_data <- weight_data[wn_neg_idx, , drop = FALSE]
    for (i in seq_len(nrow(wn_neg_data))) {
      wn_neg_data[i, ] <- scales::rescale(wn_neg_data[i, ], to = c(0.01, 100))
    }
    wn_neg_thresholds <- c(
      Matrix::rowSums(wn_neg_data) *
      (1 - abs(weight_settings$factor[wn_neg_idx] / 100))
    )
  }
```

If we set a weight to `-100` the object `wn_neg_thresholds` gets set to zero as a result of the calculations.

If we now go to line 391 we see

```
    ### add linear constraints if needed
    if (length(wn_neg_idx) > 0) {
      for (i in seq_len(nrow(wn_neg_data))) {
        initial_problem <- prioritizr::add_linear_constraints(
          initial_problem,
          threshold = wn_neg_thresholds[i],
          sense = "<=",
          data = wn_neg_data[i, initial_pu_idx]
        )
      }
    }
```

the `threshold` component gets a value of 0 meaning that for the weight we choose a value of `-100` for we are allowed to capture a max of `0` in the solution. This leads to 0 planning units selected of course which causes the error we see.

The question now is whether this should be allowed behavior of the app and lead to an informative error message or whether choosing a `weight` value of `-100` should still need to results. I think the error message route might make the most sense but am curious to learn what others think.

**Comment by DanWismer** [comment]
Thanks for taking another look at this Richard!

**Comment by DanWismer** [comment]
@ricschuster 

After chatting with @jeffreyhanson the other day, we decided that an error message is best. What I gathered from Jeff was that  if there are multiple weights enabled, it is quite computationally difficult to pin-point which weight(s) is causing the solution to be infeasible. However, once Gurobi is in use, there are some functions that can help tease that out for a more informative error message.
